From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date: Tue, 28 Feb 2023 12:12:48 -0700
Subject: [PATCH] Changes to include

- Add transitiveInclude configuration (self-explanatory)
- Don't disable transitive deps on include - the way included
  jars are resolved means we don't need to, and doing so
  breaks platform dependencies needlessly.

diff --git a/src/main/java/net/fabricmc/loom/build/nesting/IncludedJarFactory.java b/src/main/java/net/fabricmc/loom/build/nesting/IncludedJarFactory.java
index 41431b65b3ab9b229fb3ce5a96cefc49cf25d8c5..533d4eade71d05e1146e5fd201ce2cb5bd92d257 100644
--- a/src/main/java/net/fabricmc/loom/build/nesting/IncludedJarFactory.java
+++ b/src/main/java/net/fabricmc/loom/build/nesting/IncludedJarFactory.java
@@ -34,6 +34,7 @@ import java.util.Collection;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Locale;
+import java.util.Map;
 import java.util.Set;
 import java.util.function.Supplier;
 
@@ -49,7 +50,11 @@ import org.gradle.api.artifacts.ProjectDependency;
 import org.gradle.api.artifacts.ResolvedArtifact;
 import org.gradle.api.artifacts.ResolvedConfiguration;
 import org.gradle.api.artifacts.ResolvedDependency;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
+import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
+import org.gradle.api.artifacts.component.ProjectComponentIdentifier;
 import org.gradle.api.file.ConfigurableFileCollection;
+import org.gradle.api.internal.artifacts.repositories.resolver.MavenUniqueSnapshotComponentIdentifier;
 import org.gradle.api.plugins.JavaPlugin;
 import org.gradle.api.provider.Provider;
 import org.gradle.api.tasks.TaskDependency;
@@ -59,6 +64,7 @@ import org.jetbrains.annotations.Nullable;
 import net.fabricmc.loom.LoomGradleExtension;
 import net.fabricmc.loom.LoomGradlePlugin;
 import net.fabricmc.loom.task.RemapTaskConfiguration;
+import net.fabricmc.loom.util.Constants;
 import net.fabricmc.loom.util.Pair;
 import net.fabricmc.loom.util.ZipReprocessorUtil;
 import net.fabricmc.loom.util.fmj.FabricModJsonFactory;
@@ -83,6 +89,74 @@ public final class IncludedJarFactory {
 		});
 	}
 
+	public Provider<ConfigurableFileCollection> transitiveNestedJars() {
+		return this.project.provider(() -> {
+			final ConfigurableFileCollection files = this.project.files();
+			final Set<Task> builtBy = Sets.newHashSet();
+
+			final Configuration config = this.project.getConfigurations().getByName(Constants.Configurations.TRANSITIVE_INCLUDE);
+			files.from(artifacts(config, builtBy).stream().map(LazyNestedFile::file).toArray());
+			files.builtBy(config.getBuildDependencies());
+			return files;
+		});
+	}
+
+	private List<LazyNestedFile> artifacts(final Configuration config, final Set<Task> builtBy) {
+		final Set<String> seen = new HashSet<>();
+
+		final List<LazyNestedFile> files = new ArrayList<>();
+
+		final ResolvedConfiguration resolvedConfiguration = config.getResolvedConfiguration();
+		for (final ResolvedArtifact artifact : resolvedConfiguration.getResolvedArtifacts()) {
+			final ComponentIdentifier id = artifact.getId().getComponentIdentifier();
+			if (id instanceof ModuleComponentIdentifier module) {
+				final String version;
+				if (module instanceof MavenUniqueSnapshotComponentIdentifier snapshot) {
+					// vanilla loom doesn't use the timestamp
+					version = snapshot.getSnapshotVersion();
+				} else {
+					version = module.getVersion();
+				}
+				if (!seen.add(module.getGroup() + ":" + module.getModule() + ":" + version)) {
+					continue;
+				}
+				final Metadata metadata = new Metadata(
+						module.getGroup(),
+						module.getModule(),
+						version,
+						artifact.getClassifier()
+				);
+
+				files.add(new LazyNestedFile(
+						this.project,
+						metadata,
+						() -> this.getNestableJar(artifact.getFile(), metadata)
+				));
+			} else if (id instanceof ProjectComponentIdentifier projectId) {
+				final ProjectDependency dependency = (ProjectDependency) this.project.getDependencies().project(Map.of("path", projectId.getProjectPath()));
+
+				this.addProjectDependency(files, dependency, seen, builtBy);
+			}
+		}
+
+		return files;
+	}
+
+	public Provider<Pair<List<LazyNestedFile>, TaskDependency>> forgeTransitiveNestedJars() {
+		return this.project.provider(() -> {
+            final Set<Task> builtBy = Sets.newHashSet();
+
+			final Configuration config = this.project.getConfigurations().getByName(Constants.Configurations.TRANSITIVE_INCLUDE);
+
+			return new Pair<>(artifacts(config, builtBy), task -> {
+				TaskDependency dependencies = config.getBuildDependencies();
+				Set<Task> tasks = new HashSet<>(dependencies.getDependencies(task));
+				tasks.addAll(builtBy);
+				return tasks;
+			});
+		});
+	}
+
 	public Provider<Pair<List<LazyNestedFile>, TaskDependency>> getForgeNestedJars(final Configuration configuration) {
 		return project.provider(() -> {
 			final List<LazyNestedFile> files = new ArrayList<>();
@@ -131,8 +205,16 @@ public final class IncludedJarFactory {
 
 		for (Dependency dependency : configuration.getDependencies()) {
 			if (dependency instanceof ProjectDependency projectDependency) {
-				if (!visited.add(dependency.getGroup() + ":" + dependency.getName() + ":" + dependency.getVersion())) {
-					continue;
+				this.addProjectDependency(files, projectDependency, visited, builtBy);
+			}
+		}
+
+		return files;
+	}
+
+	private void addProjectDependency(final List<LazyNestedFile> files, final ProjectDependency projectDependency, final Set<String> visited, final Set<Task> builtBy) {
+				if (!visited.add(projectDependency.getGroup() + ":" + projectDependency.getName() + ":" + projectDependency.getVersion())) {
+					return;
 				}
 
 				// Get the outputs of the project
@@ -161,10 +243,6 @@ public final class IncludedJarFactory {
 						throw new UnsupportedOperationException("Cannot nest none AbstractArchiveTask task: " + task.getName());
 					}
 				}
-			}
-		}
-
-		return files;
 	}
 
 	private File getNestableJar(final File input, final Metadata metadata) {
diff --git a/src/main/java/net/fabricmc/loom/configuration/LoomConfigurations.java b/src/main/java/net/fabricmc/loom/configuration/LoomConfigurations.java
index 238a242eaaac092e22f9cb9f8e66e9c82b0d4ddd..a684a3bdada096cc526da6befd678eeceddd937e 100644
--- a/src/main/java/net/fabricmc/loom/configuration/LoomConfigurations.java
+++ b/src/main/java/net/fabricmc/loom/configuration/LoomConfigurations.java
@@ -85,6 +85,7 @@ public abstract class LoomConfigurations implements Runnable {
 		registerNonTransitive(Constants.Configurations.MINECRAFT, Role.NONE);
 		// We don't need to make this non-transitive due to the way we resolve it. Also, doing so would break platform dependencies.
 		register(Constants.Configurations.INCLUDE, Role.RESOLVABLE);
+		register(Constants.Configurations.TRANSITIVE_INCLUDE, Role.RESOLVABLE);
 		registerNonTransitive(Constants.Configurations.MAPPING_CONSTANTS, Role.RESOLVABLE);
 
 		register(Constants.Configurations.NAMED_ELEMENTS, Role.CONSUMABLE).configure(configuration -> {
diff --git a/src/main/java/net/fabricmc/loom/task/RemapJarTask.java b/src/main/java/net/fabricmc/loom/task/RemapJarTask.java
index a62dcc995b6e9f0410b04b48a1773c3a28aa30ca..9ce2716228684fd0008410861b6fd64cc6680ab8 100644
--- a/src/main/java/net/fabricmc/loom/task/RemapJarTask.java
+++ b/src/main/java/net/fabricmc/loom/task/RemapJarTask.java
@@ -153,14 +153,23 @@ public abstract class RemapJarTask extends AbstractRemapJarTask {
 
 		if (!LoomGradleExtension.get(getProject()).isForgeLike()) {
 			getNestedJars().from(factory.getNestedJars(includeConfiguration));
+			getNestedJars().from(factory.transitiveNestedJars());
 		} else {
 			Provider<Pair<List<LazyNestedFile>, TaskDependency>> forgeNestedJars = factory.getForgeNestedJars(includeConfiguration);
-			getForgeNestedJars().value(forgeNestedJars.map(Pair::left).map(pairs -> {
+			getForgeNestedJars().addAll(forgeNestedJars.map(Pair::left).map(pairs -> {
 				return pairs.stream()
 						.map(LazyNestedFile::resolve)
 						.toList();
 			}));
 			getNestedJars().builtBy(forgeNestedJars.map(Pair::right));
+
+			Provider<Pair<List<LazyNestedFile>, TaskDependency>> forgeTransitiveNestedJars = factory.forgeTransitiveNestedJars();
+			getForgeNestedJars().addAll(forgeTransitiveNestedJars.map(Pair::left).map(pairs -> {
+				return pairs.stream()
+						.map(LazyNestedFile::resolve)
+						.toList();
+			}));
+			getNestedJars().builtBy(forgeTransitiveNestedJars.map(Pair::right));
 		}
 
 		getUseMixinAP().set(LoomGradleExtension.get(getProject()).getMixin().getUseLegacyMixinAp());
diff --git a/src/main/java/net/fabricmc/loom/util/Constants.java b/src/main/java/net/fabricmc/loom/util/Constants.java
index c646772a6036908a34bba1276e8d9579127e0eb6..6ea19d4d0f6b4567ac03daf831038b47cd8ca580 100644
--- a/src/main/java/net/fabricmc/loom/util/Constants.java
+++ b/src/main/java/net/fabricmc/loom/util/Constants.java
@@ -48,6 +48,7 @@ public class Constants {
 		public static final String MOD_COMPILE_CLASSPATH = "modCompileClasspath";
 		public static final String MOD_COMPILE_CLASSPATH_MAPPED = "modCompileClasspathMapped";
 		public static final String INCLUDE = "include";
+		public static final String TRANSITIVE_INCLUDE = "transitiveInclude";
 		public static final String MINECRAFT = "minecraft";
 
 		public static final String MINECRAFT_COMPILE_LIBRARIES = "minecraftLibraries";
